### PR TITLE
MAINT: fix mypy type output

### LINKED ISSUE #2798: Type hint for page.images needs adjusting
When running mypy for type checking, if you access things like `page.images.keys()` you get a type error:

     error: "list[ImageFile]" has no attribute "keys"  [attr-defined]

This is due to the fact the return type of that property is defined as `List[ImageFile]`

https://github.com/py-pdf/pypdf/blob/cf7fcfd568bb96bb2a3b978a0bd031a18e6d90b7/pypdf/_page.py#L498

In fact the returned object (while it does imitate a sequence) is a `_VirtualListImages`

https://github.com/py-pdf/pypdf/blob/cf7fcfd568bb96bb2a3b978a0bd031a18e6d90b7/pypdf/_page.py#L535

The type hint for the image property doesn't know about this, and so when using the extra methods provided by `_VirtualListImages` such as `keys()` you hit the type checking error.

### PR DESCRIPTION
**Pull request by pubpub-zz**
closes #2798

### CODE REVIEW

**Comment by stefan6419846** [comment]
I am not completely sure about this - using private members in the public API does not feel right.

**Comment by pubpub-zz** [comment]
> I am not completely sure about this - using private members in the public API does not feel right.

Can you please clarify: keys and items are public for _VirtualListImages.
if the issue is _VirtualListImages, should we rename it into VirtualListImages. this may have more impact and will make _VirtualListImages too visible without any advantage:  I would not like to create a new subject of discussion such as the one we had on filters/LZWDecode

However, having a library fully compatible with mypy worth the commit

**Comment by stefan6419846** [comment]
In my opinion, it does not make sense to have a type hint referencing `_VirtualListImages` while this class is both not exposed publicly and even has a leading underscore to indicate it is private - such cases should only occur for the internal API. (`ImageFile` never being available as a public class is a similar issue in my opinion, but at least not prefixed by an underscore - the general issue is the same.)

**Comment by pubpub-zz** [comment]
Based on your comments: 
* I renamed `_VirtualListImages` to  `VirtualListImages` 
* I moved `ImageFile` to make it public. 
* I broke the inheritance between `ImageFile` with `File` and made it obsolete(never used out of ImageFile) in order to have a standalone module
* I also moved PIL import in _page to global in order to use `(PIL.Image.)Image` type

**Comment by stefan6419846** [comment]
> I also moved PIL import in _page to global in order to use `(PIL.Image.)Image` type

Isn't this a breaking change for users which do not have *Pillow* installed?

**Comment by pubpub-zz** [comment]
> > I also moved PIL import in _page to global in order to use `(PIL.Image.)Image` type
> 
> Isn't this a breaking change for users which do not have _Pillow_ installed?

No: the error is reported only when calling replace (as it used to be) There is no change in behavior

File: pypdf/_page.py [hunk 1]
```diff
@@ -301,6 +310,160 @@ def apply_on(
...
+class ImageFile:
+    """
+    Image within the PDF file. *This object is not designed to be built.*
+
+    This object should not be modified except using :func:`ImageFile.replace` to replace the image with a new one.
+    """
+
+    name: str = ""
+    """
+    FileName as identified within the PDF file
```
Comment (stefan6419846) [inline pypdf/_page.py:323]:
```suggestion
    Filename as identified within the PDF file.
```

File: pypdf/_page.py [hunk 2]
```diff
@@ -301,6 +310,160 @@ def apply_on(
...
+    """
+
+    data: bytes = b""
+    """
+    data as bytes
```
Comment (stefan6419846) [inline pypdf/_page.py:328]:
```suggestion
    Data as bytes.
```

File: pypdf/_page.py [hunk 3]
```diff
@@ -301,6 +310,160 @@ def apply_on(
...
+    """
+
+    image: Optional[Image] = None
+    """
+    data as PIL image;
```
Comment (stefan6419846) [inline pypdf/_page.py:333]:
```suggestion
    Data as PIL image.
```

File: pypdf/_page.py [hunk 4]
```diff
@@ -301,6 +310,160 @@ def apply_on(
...
+    """
+
+    indirect_reference: Optional[IndirectObject] = None
+    """
+    Reference to the Object storing the stream
```
Comment (stefan6419846) [inline pypdf/_page.py:338]:
```suggestion
    Reference to the object storing the stream.
```

File: pypdf/_page.py [hunk 5]
```diff
@@ -301,6 +310,160 @@ def apply_on(
...
+    """
+
+    def replace(self, new_image: Image, **kwargs: Any) -> None:
+        """
+        Replace the Image with a new PIL image.
```
Comment (stefan6419846) [inline pypdf/_page.py:343]:
```suggestion
        Replace the image with a new PIL image.
```

File: pypdf/_page.py [hunk 6]
```diff
@@ -301,6 +310,160 @@ def apply_on(
...
+            )
+
+        from ._reader import PdfReader
+
+        # to prevent circular import
+        from .filters import _xobj_to_image
+        from .generic import DictionaryObject, PdfObject
+
+        if self.indirect_reference is None:
+            raise TypeError("Can not update an inline image")
```
Comment (stefan6419846) [inline pypdf/_page.py:373]:
```suggestion
            raise TypeError("Cannot update an inline image.")
```

File: pypdf/_page.py [hunk 7]
```diff
@@ -301,6 +310,160 @@ def apply_on(
...
+        if not hasattr(self.indirect_reference.pdf, "_id_translated"):
+            raise TypeError("Can not update an image not belonging to a PdfWriter")
```
Comment (stefan6419846) [inline pypdf/_page.py:375]:
```suggestion
            raise TypeError("Cannot update an image not belonging to a PdfWriter.")
```

File: pypdf/_page.py [hunk 8]
```diff
@@ -301,6 +310,160 @@ def apply_on(
...
+        return f"{self.__class__.__name__}(name={self.name}, data: {_human_readable_bytes(len(self.data))})"
+
+    def __repr__(self) -> str:
+        return self.__str__()[:-1] + f", hash: {hash(self.data)})"
+
+
+class VirtualListImages(Sequence[ImageFile]):
+    """
+    Provides access to images referenced within a page.
+    One copy only will be returned if used many times in the same page
```
Comment (stefan6419846) [inline pypdf/_page.py:407]:
```suggestion
    Only one copy will be returned if the usage is used on the same page multiple times.
```

File: pypdf/_page.py [hunk 9]
```diff
@@ -301,6 +310,160 @@ def apply_on(
...
+    see :func:`PageObject.images` for more details
```
Comment (stefan6419846) [inline pypdf/_page.py:408]:
```suggestion
    See :func:`PageObject.images` for more details.
```

File: pypdf/_utils.py [hunk 10]
```diff
@@ -525,10 +524,18 @@ def getter(self, method):  # type: ignore  # noqa: ANN001, ANN202
 class File:
     from .generic import IndirectObject
 
-    name: str
-    data: bytes
-    image: Optional[Any] = None  # optional ; direct image access
-    indirect_reference: Optional[IndirectObject] = None  # optional ; link to PdfObject
+    name: str = ""
+    """
+    FileName as identified within the PDF file
```
Comment (stefan6419846) [inline pypdf/_utils.py:529]:
```suggestion
    Filename as identified within the PDF file.
```

File: pypdf/_utils.py [hunk 11]
```diff
@@ -525,10 +524,18 @@ def getter(self, method):  # type: ignore  # noqa: ANN001, ANN202
...
+    """
+    data: bytes = b""
+    """
+    data as bytes
```
Comment (stefan6419846) [inline pypdf/_utils.py:533]:
```suggestion
    Data as bytes.
```

File: pypdf/_utils.py [hunk 12]
```diff
@@ -525,10 +524,18 @@ def getter(self, method):  # type: ignore  # noqa: ANN001, ANN202
...
+    """
+    indirect_reference: Optional[IndirectObject] = None
+    """
+    Reference to the Object storing the stream
```
Comment (stefan6419846) [inline pypdf/_utils.py:537]:
```suggestion
    Reference to the object storing the stream.
```

File: tests/test_workflows.py [hunk 13]
```diff
@@ -990,6 +990,13 @@ def test_replace_image(tmp_path):
         i.replace(reader.pages[0].images[0].image)
     assert exc.value.args[0] == "Can not update an inline image"
 
+    import pypdf
+
+    pypdf._page.pil_not_imported = True
```
Comment (stefan6419846) [inline tests/test_workflows.py:995]: This should use a construct with `finally` to ensure that we always reset the value correctly.

**Review by stefan6419846 — COMMENTED** [review]
I have reworded some of the docs.

Additionally, the `ImageFile` and `VirtualListImages` classes are still part of a private module only without being exposed in the `__init__.py` file. We should expose them there to allow users to actually refer to them in their own type hints if required.

**Comment by pubpub-zz** [comment]
> Additionally, the `ImageFile` and `VirtualListImages` classes are still part of a private module only without being exposed in the `__init__.py` file. We should expose them there to allow users to actually refer to them in their own type hints if required.

ImageFile and VirtualListImages are not classes that a user should create instances from : as such I see no reason for them to be part of the `__init__.py`

The objective is just to fix mypy and the test report in the issue was positive

**Comment by stefan6419846** [comment]
Okay, let's keep this for now - except for the other remark regarding `finally` which still should be fixed.

**Review by stefan6419846 — APPROVED** [review]

**Pull request merged**